### PR TITLE
Mutualize audio readings

### PR DESCRIPTION
**Pull request by artemru**
**What does this PR do? Please describe:**
- This PR mutualizes the code between different Speech and ASR dataloaders with static method and subclassing
 - It also introduces parquet speech and asr dataloaders family (to load audio directly from binary inside parquets)
 - It enables batch and example level dataset (static) weighted mixture for all ASR dataloaders (it can be next generized on all other type of dataloading)

Minor points :
 - using `pa.jemalloc_memory_pool()` for faster memory release when loading parquet
 - using `SpeechReadOptions` for both Asr and Speech datasets

Some other important behavior changes:
- for train splits (detected with "train" in split) we do infinite loop in datapipeline to avoid resets
- this is also coherent with mixture type dataloading (infinite loop)
- for non-train splits, we do one epoch only

## Tests
- reaching parity on ASR finetuning metrics
- currently running pretraining (there's some different behavior in diversity loss, -> i'll play more with shuffling window)
- running some mixture recipes to verify the integration

## Config considerations

- To enable data mixture one needs to use the following pattern the preset:
```python
        config.dataset.name = "dataset1:7,dataset2:1"
        config.dataset.family = "weighted_dataset_mixture"
        config.dataset.train_split = "dataset1=train,dataset2=train"
        config.dataset.valid_split = "dataset1=[dev,test],dataset2=[dev,test]"
        config.dataset.extras = {"mixing_level": "batch"}  # {"mixing_level": "example"}
```
- For parquet dataloader, several params are passed through `config.extras`, in particular :
    - fragment_shuffle_window ( default = -1, global fragments shuffling)
    - columns (possibility to pass custom column config)
    - pa_cpu_count (default to 20)
    - partition_filters (default to None)
    - files_circular_shift (default to False)

- For parquet dataloader, the examples and batch shuffling happens in memory, having very large window will result in higher cpu memory footprint. As dummy protection mechanics, we added upper some reasonable boundaries and not allow global window=0 any more.

### CODE REVIEW

File: src/fairseq2/datasets/speech_parquet.py [hunk 1]
```diff
@@ -75,19 +59,15 @@ class DefaultAudioSchema(NamedColumns):
...
     _dataset: pq.ParquetDataset
     _splits: set[str]
     split_column: str = "split"
 
-    nb_samples_per_fragment = 1000  # FIXME: detect it from the dataset metadata
-    max_num_batches: int = 1000
-    max_num_examples: int = 2_000_000
-    pa_cpu_count: int = 20
+    max_num_batches: int = 2000
+    max_num_examples: int = 100_000
```
Comment (mattsetz) [inline src/fairseq2/datasets/speech_parquet.py:70]: seems low ... did we copy these defaults over from the old dataset implementations?

File: src/fairseq2/datasets/speech_parquet.py [hunk 2]
```diff
@@ -38,32 +37,17 @@
...
-def rename_feature(batch: List[Dict[str, Any]]) -> List[Dict[str, Any]]:
-    for example in batch:
-        if "fbank" in example["audio"]:
-            example["audio_feature"] = example["audio"].pop("fbank")
-        elif "waveform" in example["audio"]:
-            example["audio_feature"] = example["audio"].pop("waveform")
-    return batch
-
-
 PARQUET_SPEECH_DATASET_FAMILY: Final = "generic_parquet_speech"
```
Comment (mattsetz) [inline src/fairseq2/datasets/speech_parquet.py:53]: nit: maybe `audio_length` is more clear?

File: src/fairseq2/datasets/speech_parquet.py [hunk 3]
```diff
@@ -100,7 +80,7 @@ def from_path(
         path: Path | str | List[str | Path],
         name: str,
         filesystem: Any | None = None,
-    ) -> GenericSpeechParquetDataset:
+    ) -> "ParquetDatasetInterface":
 
         # from stopes.fb_config import get_filesystem_from_path
         # path, filesystem = get_filesystem_from_path(path)  # type: ignore
```
Comment (mattsetz) [inline src/fairseq2/datasets/speech_parquet.py:86]: nit: line 87 `datasest` --> `dataset`

**Review by mattsetz — APPROVED** [review]

File: src/fairseq2/datasets/asr.py [hunk 4]
```diff
@@ -92,47 +75,33 @@ def splits(self) -> set[str]:
         """Return the set of splits."""
 
 
-# TODO: FIX, INFER
-npc = 10
-
-
 GENERIC_ASR_DATASET_FAMILY: Final = "generic_asr"
 
+get_asr_dataset_hub = DatasetHubAccessor(AsrDataset)
```
[the remark this answers is not in the document]
Comment (artemru) [reply src/fairseq2/datasets/asr.py:80]: it's more of an utility function if one wants to load a dataset through Hub

File: src/fairseq2/datasets/asr_parquet.py [hunk 5]
```diff
@@ -0,0 +1,155 @@
...
+from fairseq2.models.seq2seq import Seq2SeqBatch
+
+PARQUET_ASR_DATASET_FAMILY: Final = "generic_parquet_asr"
+
+
+@dataclass
+class DefaultASRSchema(NamedColumns):
+    audio: str = "audio_bytes"
+    length: str = "audio_size"
+    text: str = "text"
```
[the remark this answers is not in the document]
Comment (artemru) [reply src/fairseq2/datasets/asr_parquet.py:36]: yes, good point actually ! let me change it!

File: src/fairseq2/datasets/asr_parquet.py [hunk 6]
```diff
@@ -0,0 +1,155 @@
...
+        if split not in self._splits:
+            raise UnknownSplitError(self._name, split, self._splits)
+
+        if options is None:
+            options = SpeechReadOptions()
+
+        seed = options.seed
+
+        options.batch_shuffle_window = min(
+            options.batch_shuffle_window, self.max_num_batches
```
[the remark this answers is not in the document]
Comment (artemru) [reply src/fairseq2/datasets/asr_parquet.py:62]: yeah, but here we dont want to do that (and I can put some protection against it) since we dont want to shuffle full dataset (with audio included) in memory. Note that the shuffling here will be somehow global since we shuffle parquet fragments (globally) at the beginning of the pipeline, the example level shuffling is only useful to mix up together the elements from different fragments.

File: src/fairseq2/datasets/asr_parquet.py (hunk 5 above)
[the remark this answers is not in the document]
Comment (artemru) [reply src/fairseq2/datasets/asr_parquet.py:36]: oh, actually, we'll be able to change it later -> right now it's "hard-coded" name in manifests ...

File: src/fairseq2/datasets/speech.py [hunk 7]
```diff
@@ -189,14 +161,12 @@ def __init__(self, name: str, manifest_dir: Path, splits: set[str]) -> None:
         self._manifest_dir = manifest_dir
         self._splits = splits
 
-    @staticmethod
-    def from_path(path: Path, name: str) -> GenericSpeechDataset:
+    @classmethod
+    def from_path(cls, path: Path, name: str) -> "ManifestDatasetInterface":
```
[the remark this answers is not in the document]
Comment (artemru) [reply src/fairseq2/datasets/speech.py:165]: it should be a classmethod, so it can return self of the correct class (remember the issue we've got with staticmethod when we tried to subclass it)

File: src/fairseq2/datasets/speech_parquet.py (hunk 1 above)
Comment (artemru) [reply src/fairseq2/datasets/speech_parquet.py:59]: they does not have exactly the same semantic since as before (since all small fragments are shuffled globally).
Right now, it's more a memory tradeoff, since this example and batch shuffling will happen in memory which requires to hold all encoded audio being in memory.

I will increase those bounds even more (with this values we consume less <80G per gpus, so we have some extra room here).

File: src/fairseq2/datasets/speech_parquet.py (hunk 2 above)
Comment (artemru) [reply src/fairseq2/datasets/speech_parquet.py:42]: the idea is that we can ingest datasets with different columns names through the NamedColumns config.
In next PR we can expose this column names in the datacards!
